### PR TITLE
chessformer

### PR DESCRIPTION
**Pull request by lichensongs**

### CODE REVIEW

File: py/shared/chessformer_modules.py [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+            logits = (logits
+                      + torch.einsum('bhtd,htsd->bhts', Q, rpe_k) * self.scale
+                      + torch.einsum('htsd,bhsd->bhts', rpe_q, K) * self.scale)
+
+        # Smolgen dynamic bias
+        if self.use_smolgen:
+            logits = logits + self.smolgen(x)  # (B,Hh,T,T)
+
+        attn = torch.softmax(logits, dim=-1)
+        attn = self.dropout(attn)
```
Comment (shindavid) [inline py/shared/chessformer_modules.py:181]: This dropout is interesting. Is this to match LC0 Chessformer?

Should we be using dropout in other modules of ours?

Does the relative ordering of softmax and dropout matter?

File: py/shared/chessformer_modules.py [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+
+        out = torch.matmul(attn, V)  # (B,Hh,T,dh)
+
+        if self.use_rpe:
+            rpe_v = self._expand_rpe_logits(self.rpe_v)  # (Hh, dh, T, T)
+            out = out + torch.einsum('bhts,htsd->bhtd', attn, rpe_v)
+
+        out = out.transpose(1, 2).contiguous().view(B, T, Dm)  # (B,T,Dm)
+        out = self.dropout(self.Wo(out)) * self.res_scale
```
Comment (shindavid) [inline py/shared/chessformer_modules.py:190]: Is it correct to dropout twice? Does this match LC0 Chessformer?

File: py/shared/chessformer_modules.py [hunk 3]
```diff
@@ -0,0 +1,372 @@
+import math
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+from typing import Optional
+
+class Mish(nn.Module):
```
Comment (shindavid) [inline py/shared/chessformer_modules.py:8]: Is "mish" a standard term? I never heard of it.

File: py/shared/chessformer_modules.py [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+        if self.use_rpe:
+            rpe_v = self._expand_rpe_logits(self.rpe_v)  # (Hh, dh, T, T)
+            out = out + torch.einsum('bhts,htsd->bhtd', attn, rpe_v)
+
+        out = out.transpose(1, 2).contiguous().view(B, T, Dm)  # (B,T,Dm)
+        out = self.dropout(self.Wo(out)) * self.res_scale
+        return out
+
+
+class FFN(nn.Module):
```
Comment (shindavid) [inline py/shared/chessformer_modules.py:194]: I really prefer using more descriptive names, and avoiding acronyms, especially in class names. It's not obvious to me what this FFN stands for.

File: py/shared/chessformer_modules.py [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+        y = self.fc2(self.act(self.fc1(x)))
+        return self.dropout(y) * self.res_scale
+
+
+class EncoderLayer(nn.Module):
+    def __init__(self, Dm: int, Hh: int, T: int, Dff: int, n_layers: int,
+                 use_static_bias=True, use_rpe: bool = False,
+                 rpe_factorizer: Optional[torch.Tensor] = None,
+                 use_smolgen=True, smolgen_shared=None,
+                 smolgen_shared_dim: int = 256, smolgen_compress_dim: int = 32):
```
Comment (shindavid) [inline py/shared/chessformer_modules.py:216]: Anytime you have this many arguments, you should collapse into 1 or more params objects.

One option is to just throw everything into a single `EncoderParams`.

Alternatively, you could chunk it up a bit, like have `Optional[SmolgenParams]` for the `*smolgen*` values.

File: py/alphazero/servers/loop_control/params.py [hunk 6]
```diff
@@ -57,7 +57,7 @@ def add_args(parser, include_cuda_device=True):
         group.add_argument('--task-mode', action='store_true', default=defaults.task_mode,
                            help='run without self-play/training; exit after task is done.')
         group.add_argument('--retrain', action='store_true', default=False,
-                           help='retrain models from scratch, even if model files are present.')
+                           help='retrain models for a forked run without running self-eval or eval-vs-benchmark.')
```
Comment (shindavid) [inline py/alphazero/servers/loop_control/params.py:60]: "self-eval" -> "self-play"?

Maybe this arg should be renamed "train-only"? Because that's what it does: it makes it so that loop-controller does training and nothing else. Can specify in the help-str that this can only be used for a run that was forked via fork_run.py --retrain-models.

File: py/alphazero/servers/loop_control/directory_organizer.py [hunk 7]
```diff
@@ -255,27 +258,32 @@ def get_any_self_play_data_filename(self, gen: Optional[Generation]) -> Optional
...
-                            last_model_gen: Optional[Generation] = None):
+
+    def _parse_self_play_data_filename(self, filename: str) -> Gen:
+        m = SELF_PLAY_DATA_FILE_REGEX.fullmatch(filename)
+        if not m:
+            raise ValueError(f'Unexpected file in self-play-data: {filename}')
+        return int(m.group(1))
+
+    def _apply_to_self_play_data_dir(self, target: 'DirectoryOrganizer',
+                                     func: Callable[[PathStr, PathStr], None],
```
[the remark this answers is not in the document]
Comment (shindavid) [reply py/alphazero/servers/loop_control/directory_organizer.py:274]: Actually, I remember now. You can do `Callable[Args, Any]`.

File: py/games/connect4/spec.py [hunk 8]
```diff
@@ -208,7 +209,7 @@ def generate(shape_info_dict: ShapeInfoDict) -> ModelConfig:
         )
 
 
-class Chessformer(ModelConfigGenerator):
+class Transformer(ModelConfigGenerator):
```
Comment (shindavid) [inline py/games/connect4/spec.py:212]: Can you delete the existing `Transformer` class?

File: py/alphazero/scripts/run_local.py [hunk 9]
```diff
@@ -374,27 +374,29 @@ def main():
...
-                procs.append(launch_eval_vs_benchmark_server(params_dict, ratings_gpu, game_spec))
-            if params.run_ratings_server:
-                descs.append('Ratings')
-                procs.append(launch_ratings_server(params_dict, ratings_gpu))
-        else:
-            if benchmark_data.valid():
-                descs.append('Eval')
-                procs.append(launch_eval_vs_benchmark_server(params_dict, ratings_gpu, game_spec))
+
+        if not params.retrain:
```
Comment (shindavid) [inline py/alphazero/scripts/run_local.py:378]: In the `if params.retrain:` case, can you add an assert:

```
assert directory_organizer.requires_retraining()
```
